### PR TITLE
Add converter for index.Tensor_hacked_twin

### PR DESCRIPTION
**Pull request by matthewfl**
* Add converter for index.Tensor_hacked_twin -> tcp.gather & tcp.broadcast
* Remove the  `tcp.custom_op` variants of `Tensor_hacked_twin`
  * The Tcp variants do not have full coverage of the PyTorch, but we should seek to expand the coverage of our converters
* Fix tcp.const when it is a dense resource of ints to cast the value
* Add verifier for `tcp.gather` and `tcp.const` to ensure that is used correctly

### CODE REVIEW

File: test/Conversion/TorchToTcp/data_movement.mlir [hunk 1]
```diff
@@ -64,3 +64,29 @@ func.func @torch.aten.index_select(%arg0: !torch.vtensor<[4,3],f32>, %arg1: !tor
...
+// CHECK: return %[[RET]]
+func.func @torch.aten.index.tensor_hacked_twin(%arg0: !torch.vtensor<[1,20,30],f32>, %select1: !torch.vtensor<[5,1],si64>, %select2: !torch.vtensor<[20],si64>) -> !torch.vtensor<[1,5,20],f32> {
+  // there is a strange pattern that is being generated when selecting one axis.  It seems that it uses the Tensor_hacked_twin to select along all axis, but uses
+  // arange to select all of the 
+  %none = torch.constant.none
+  %int0 = torch.constant.int 0
+  %int1 = torch.constant.int 1
+  %int4 = torch.constant.int 4 // this is a dtype on arange....
+  %int-1 = torch.constant.int -1
+  %arange = torch.aten.arange.start_step %int0, %int1, %int1, %int4, %none, %none, %none : !torch.int, !torch.int, !torch.int, !torch.int, !torch.none, !torch.none, !torch.none -> !torch.vtensor<[1],si64>
```
Comment (zezhang) [inline test/Conversion/TorchToTcp/data_movement.mlir:85]: `torch.aten.arange.start_step` will remain unchanged right? As we don't handle it in tcp.

**Review by zezhang — COMMENTED** [review]
LGTM! Can we also try to add it to the [aot_compile](https://github.com/cruise-automation/mlir-tcp/blob/main/test/AotCompile/model_loader_lib.py) test suite for e2e coverage?

File: test/Conversion/TorchToTcp/data_movement.mlir (hunk 1 above)
Comment (matthewfl) [reply test/Conversion/TorchToTcp/data_movement.mlir:85]: Currently this remains unchanged.  I was thinking that we should add an optimization pass that removes this when it selects everything, but I have yet to do that

File: lib/Dialect/IR/TcpOps.cpp [hunk 2]
```diff
@@ -170,6 +176,31 @@ LogicalResult CastOp::verify() {
...
+    return emitOpError("tcp.gather requires that the input tensor and indices "
+                       "are the same rank");
+
+  for (int i = 0; i < inputTensor.getRank(); i++) {
+    if (inputTensor.getShape()[i] != indicesTensor.getShape()[i] &&
+        !(inputTensor.getShape()[i] == ShapedType::kDynamic ||
+          i == gatherDim)) {
+      return emitOpError("indices tensor does not match expected shape");
+    }
+  }
```
Comment (navahgar) [inline lib/Dialect/IR/TcpOps.cpp:194]: This doesn't seem to match the semantics of `tcp.gather`, which should have the same constraints as `torch.gather`.

Specifically the following from [here](https://pytorch.org/docs/stable/generated/torch.gather.html#torch.gather)
> It is also required that index.size(d) <= input.size(d) for all dimensions d != dim

Comment (matthewfl) [reply lib/Dialect/IR/TcpOps.cpp]: I updated it

File: lib/Dialect/IR/TcpOps.cpp [hunk 3]
```diff
@@ -170,6 +176,32 @@ LogicalResult CastOp::verify() {
...
+  if (inputTensor.getRank() != indicesTensor.getRank())
+    return emitOpError("tcp.gather requires that the input tensor and indices "
+                       "are the same rank");
+
+  for (int i = 0; i < inputTensor.getRank(); i++) {
+    if (inputTensor.getShape()[i] < indicesTensor.getShape()[i] &&
+        !(inputTensor.getShape()[i] == ShapedType::kDynamic ||
+          indicesTensor.getShape()[i] == ShapedType::kDynamic ||
+          i == gatherDim)) {
+      return emitOpError("indices tensor does not match expected shape");
```
Comment (navahgar) [inline lib/Dialect/IR/TcpOps.cpp:193]: Can you update this error to explicitly say the exact condition that fails here? That might be useful to know.

**Review by navahgar — APPROVED** [review]
LGTM for `GatherOp::verify()`

File: lib/Dialect/IR/TcpOps.cpp (hunk 3 above)
Comment (matthewfl) [reply lib/Dialect/IR/TcpOps.cpp]: done

**Review by zezhang — APPROVED** [review]
LGTM! Thanks for working on this tricky case.

**Pull request merged**